src/rust/engine/graph/src/context.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use std::ops::Deref;
+use std::sync::atomic::{self, AtomicU32, AtomicUsize};
+use std::sync::Arc;
+
+use workunit_store::RunId;
+
+use crate::node::{CompoundNode, EntryId, Node};
+use crate::Graph;
+
+struct InnerContext<N: Node + Send> {
+  context: N::Context,
+  run_id: AtomicU32,
+  stats: Stats,
+  graph: Graph<N>,
+}
+
+///
+/// A context passed between running Nodes which is used to request and record dependencies.
+///
+/// Parametrized by:
+///     N: Node - The Node type that this Context is being used for.
+///
+#[derive(Clone)]
+pub struct Context<N: Node + Send> {
+  entry_id: Option<EntryId>,
+  inner: Arc<InnerContext<N>>,
+}
+
+impl<N: Node + Send> Context<N> {
+  pub(crate) fn new(graph: Graph<N>, context: N::Context, run_id: RunId) -> Self {
+    Self {
+      entry_id: None,
+      inner: Arc::new(InnerContext {
+        context,
+        run_id: AtomicU32::new(run_id.0),
+        stats: Stats::default(),
+        graph,
+      }),
+    }
+  }
+
+  ///
+  /// Get the future value for the given Node implementation.
+  ///
+  pub async fn get<CN: CompoundNode<N>>(&self, node: CN) -> Result<CN::Item, N::Error> {
+    let node_result = self
+      .inner
+      .graph
+      .get(self.entry_id, self, node.into())
+      .await?;
+    Ok(
+      node_result
+        .try_into()
+        .unwrap_or_else(|_| panic!("A Node implementation was ambiguous.")),
```

Comment:
As is this message! Can you format in any information that would make this less bewildering if it happens?

Reply:
Yea, good idea. I think that I wrote this before I knew about `std::any::type_name`.

Reply:
And #18855 replaces the panic.